### PR TITLE
New UI and rewrite pardus-boot-repair

### PR DESCRIPTION
**Pull request by Vilez0**
This pull request rewrites the core functionality using Python3 and GTK3 for a more modern and user-friendly experience.

**Key Changes:**

**Complete Rewrite:** The program is now written in Python3 and leverages GTK3 with libhandy for a cleaner and more maintainable codebase.
**Reducing Code Duplication:** A reusable script for mounting the root filesystem is created and used throughout the program(0a04640)
**Modern UI:** A graphical user interface built with GTK3 and libhandy provides a more intuitive user experience. (50c0165)
**Btrfs Support:** The tool now supports Btrfs subvolumes for improved compatibility. (0a04640)
**Legacy Removal:** Obsolete Zenity-based code is removed in favor of the more modern approach. (1b8a304)
**Makefile Updates:** The Makefile is updated to reflect the new codebase and dependencies. (74d498e)
**Bug Fixes:** A logical error in the reset-password script is corrected. (4547b2b)
**Licensing and Documentation:** A license is added, and the README is updated with clear instructions and screenshots.(ca9a1fa 85500e0)
**Testing:**

I've thoroughly tested the rewritten Pardus Boot Repair through manual testing on various Pardus installations. All functionalities work as expected, and the new UI provides a smooth user experience.

**Additional Notes:**

This pull request represents a substantial modernization of Pardus Boot Repair. please feel free to leave comments or ask questions if you have any.

**New UI vs Old UI:**
image
(more screenshots can be found in the new readme)

**Todo**:
- [x] Update screenshots ( #4 )

### CODE REVIEW

File: src/scripts/pardus-reinstall [hunk 1]
```diff
@@ -1,17 +1,13 @@
...
-    echo -n "Rootfs partition >>> "
-    read disk
-done
 while [[ ! -b /dev/$mbr ]] ; do
     echo $(ls /sys/block/ | grep -v "loop")
     echo -n "Grub disk >>> "
     read mbr
 done
-mount /dev/$disk /target
+. pardus-mount-rootfs
```
Comment (sulincix) [inline src/scripts/pardus-reinstall]: This import path is  must full path

File: src/scripts/reset-password [hunk 2]
```diff
@@ -1,20 +1,11 @@
...
-    echo -n "Rootfs partition >>> "
-    read disk
-done
-
-# Unmount the partition if it is already mounted
-umount -lf /dev/$disk >/dev/null 2>&1 || true
 
-# Mount the partition to DESTDIR
-mount /dev/$disk "$DESTDIR"
+. pardus-mount-rootfs
```
Comment (sulincix) [inline src/scripts/reset-password]: This import path is  must full path

File: src/scripts/pardus-chroot [hunk 3]
```diff
@@ -1,30 +1,22 @@
...
 if [[ $UID -ne 0 ]] ; then
     echo "You must be root!"
     exit 1
 fi
+# $1 will be something like /dev/sda1
 if [[ -b "$1" ]] ; then
     DESTDIR=$(mktemp -d)
-    mount "$1" "$DESTDIR"
+    disk=$(echo "$1" | sed 's/\/dev\///')
+    . pardus-mount-rootfs
```
Comment (sulincix) [inline src/scripts/pardus-chroot]: This import path is  must full path

File: src/scripts/grub-reinstall [hunk 4]
```diff
@@ -1,38 +1,27 @@
...
     echo -n "Grub disk >>> "
     read mbr
 done
-umount -lf /dev/$disk 2>/dev/null  || true
-mount /dev/$disk "$DESTDIR"
-mount --bind /dev "$DESTDIR"/dev
-mount --bind /dev/pts "$DESTDIR"/dev/pts
-mount --bind /sys "$DESTDIR"/sys
-mount --bind /proc "$DESTDIR"/proc
+. pardus-mount-rootfs
```
Comment (sulincix) [inline src/scripts/grub-reinstall]: This import path is  must full path

File: src/scripts/full-upgrade [hunk 5]
```diff
@@ -1,19 +1,10 @@
...
-    echo $(ls /sys/block/* | grep "[0-9]$")
-    echo -n "Rootfs partition >>> "
-    read disk
-done
-mount /dev/$disk "$DESTDIR"
-mount --bind /dev "$DESTDIR"/dev
-mount --bind /dev/pts "$DESTDIR"/dev/pts
-mount --bind /sys "$DESTDIR"/sys
-mount --bind /proc "$DESTDIR"/proc
+. pardus-mount-rootfs
```
Comment (sulincix) [inline src/scripts/full-upgrade]: This import path is  must full path

File: src/scripts/fix-broken-packages [hunk 6]
```diff
@@ -1,18 +1,10 @@
...
-    echo -n "Rootfs partition >>> "
-    read disk
-done
-mount /dev/$disk "$DESTDIR"
-mount --bind /dev "$DESTDIR"/dev
-mount --bind /dev/pts "$DESTDIR"/dev/pts
-mount --bind /sys "$DESTDIR"/sys
-mount --bind /proc "$DESTDIR"/proc
+
+. pardus-mount-rootfs
```
Comment (sulincix) [inline src/scripts/fix-broken-packages]: This import path is  must full path

File: src/scripts/dump-info-log [hunk 7]
```diff
@@ -1,15 +1,12 @@
...
-    echo -n "Input output target"
+    echo -n "Enter the target directory to save the logs to: "
     read target
 fi
-if [[ "$disk" == "" || ! -b "/dev/$disk" ]] ; then
-    echo -n "Input rootfs disk"
-    read disk
-fi
-mount "/dev/$disk" "$DESTDIR"
+. pardus-mount-rootfs
```
Comment (sulincix) [inline src/scripts/dump-info-log]: This import path is  must full path

File: src/scripts/pardus-mount-rootfs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        umount "$DESTDIR" -f -l
+        mount -o subvol="$subvolume" /dev/$disk "$DESTDIR"
+    fi
+fi
+
+for path in dev dev/pts proc sys; do
+    mount --bind "/$path" "$DESTDIR/$path"
+done
+
+if [[ $fs == 'btrfs' ]]; then
```
Comment (sulincix) [inline src/scripts/pardus-mount-rootfs]: I think this check is unneeded.

File: src/scripts/pardus-mount-rootfs [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    echo -n "Rootfs partition >>> "
+    read disk
+done
+
+umount -lf /dev/$disk 2>/dev/null  || true
+fs="$(lsblk -no FSTYPE /dev/$disk | tr -d '\n')"
+mount /dev/$disk "$DESTDIR"
+
+# btrfs subvolumes
+if [[ $fs == 'btrfs' ]]; then
```
Comment (sulincix) [inline src/scripts/pardus-mount-rootfs:30]: You can merge subvolume mounth with fstab mount

File: src/scripts/pardus-reinstall (hunk 1 above)
Comment (Vilez0) [reply src/scripts/pardus-reinstall]: done

File: src/scripts/grub-reinstall (hunk 4 above)
Comment (Vilez0) [reply src/scripts/grub-reinstall]: done

File: src/scripts/reset-password (hunk 2 above)
Comment (Vilez0) [reply src/scripts/reset-password]: done 1b07c46

File: src/scripts/pardus-chroot (hunk 3 above)
Comment (Vilez0) [reply src/scripts/pardus-chroot]: done 1b07c46

File: src/scripts/full-upgrade (hunk 5 above)
Comment (Vilez0) [reply src/scripts/full-upgrade]: done 1b07c46

File: src/scripts/fix-broken-packages (hunk 6 above)
Comment (Vilez0) [reply src/scripts/fix-broken-packages]: done 1b07c46

File: src/scripts/dump-info-log (hunk 7 above)
Comment (Vilez0) [reply src/scripts/dump-info-log]: done 1b07c46

File: src/scripts/pardus-mount-rootfs (hunk 8 above)
Comment (Vilez0) [reply src/scripts/pardus-mount-rootfs]: done f29579e

File: src/Main.py [hunk 10]
```diff
@@ -0,0 +1,636 @@
...
+            if self.get_rootfs(widget) == None:
+                return
+            if self.get_mbr(widget) == None:
+                return
+            if os.path.exists("/sys/firmware/efi/efivars") and self.get_clearEfivars(widget) == None: 
+                return
+            self.update_status_page(_("Reinstalling GRUB Bootloader"), "content-loading-symbolic", _("We're reinstalling the GRUB boot loader to ensure your system can start up properly. This process may take a few moments. Once complete, your computer should boot into Pardus as usual."), False, False)
+            self.post_command = post
+            if self.rootfs.root_subvol == None:
+                self.vte_command("env disk={} mbr={} pardus-reinstall".format(self.rootfs.name, self.mbr))
```
Comment (sulincix) [inline src/Main.py]: trigger grub-reinstall instead of pardus-reinstall

File: src/Main.py [hunk 11]
```diff
@@ -0,0 +1,636 @@
...
+            else:
+                self.vte_command("env subvolume={} disk={} mbr={} grub-reinstall".format(self.rootfs.root_subvol, self.rootfs.name, self.mbr))
+        def post(Terminal, widget):
```
Comment (sulincix) [inline src/Main.py]: check script status. If failed show fail message.

File: src/Main.py [hunk 12]
```diff
@@ -0,0 +1,636 @@
...
+class Application(Gtk.Application):
+    def __init__(self, *args, **kwargs):
+        super().__init__(
+          *args, 
+          application_id="tr.org.pardus.boot-repair",
+          **kwargs
+        )
+
+        # main
+        self.builder = Gtk.Builder()
```
Comment (sulincix) [inline src/Main.py:35]: headerbar window controls are missing.

Comment (sulincix) [reply src/Main.py:35]: We can add show logs button on headerbar. 
You can add about dialog.

Pardus footer missing.

Comment (sulincix) [reply src/Main.py:35]: you must clear log screen before new operations.

Comment (sulincix) [reply src/Main.py:35]: You can add operating system name in partition selection page. (using search-operating-system script)

File: src/Main.py (hunk 10 above)
Comment (Vilez0) [reply src/Main.py]: done

File: src/Main.py (hunk 11 above)
Comment (Vilez0) [reply src/Main.py]: Done

File: src/Main.py [hunk 13]
```diff
@@ -0,0 +1,678 @@
...
+            users = self.list_users(self.rootfs)
+            if len(users) == 0:
+                self.update_status_page(_("No users found"), "dialog-error-symbolic", _("No users found"), True, True)
+                return
+            self.update_status_page(_("Resetting User Settings"), "content-loading-symbolic", _("We're resetting your user configuration to its default state. This will revert any custom settings back to their original values. Please note that any personalized preferences will be lost. Once complete, your system will be refreshed and ready for use."), False, False)
+            self.post_command = post
+            if self.rootfs.root_subvol == None:
+                self.vte_command("env pardus-chroot /dev/{} su {} -c 'cd ; rm -rvf .dbus .cache .local .config'".format(self.rootfs.name ,users[0]))
+            else:
+                self.vte_command("env subvolume={} pardus-chroot /dev/{} su {} -c 'cd ; rm -rvf .dbus .cache .local .config'".format(self.rootfs.root_subvol, self.rootfs.name ,users[0]))
```
Comment (sulincix) [inline src/Main.py]: Ask for user selection. System can have multiple user.

File: src/Main.py [hunk 14]
```diff
@@ -0,0 +1,678 @@
...
+    def update_status_page(self, title, icon_name, description, stop_spinner=False, enable_mainpage=True):
+        self.status_page.set_title(title)
+        self.status_page.set_icon_name(icon_name)
+        self.status_page.set_description(description)
+        self.btn_go_mainpage.set_sensitive(enable_mainpage)
+        self.spinner_loading.start()
+        if stop_spinner:
+            self.spinner_loading.stop()
+
+    def vte_command(self, command):
```
Comment (sulincix) [inline src/Main.py]: You must ask question before action. Action reasons can be data loss or other bad stuff.

File: src/Main.py [hunk 15]
```diff
@@ -0,0 +1,678 @@
...
+        except Exception as e:
+            # write error to stderr
+            sys.stderr.write(str(e) + "\n")
+            self.update_status_page(_("An error occured"), "dialog-error-symbolic", str(e), True, True)
+
+    def vte_cb(self, pid, error):
+        if error != None or pid == -1:
+            self.update_status_page(_("An error occured"), "dialog-error-symbolic", _("An error occurred before the command has been executed"), True, True)
+            return
+    def vte_exited(self, widget, status):
```
Comment (sulincix) [inline src/Main.py]: Run chroot shell
press ctrl + d

Run shell again
press ctrl + d

application crash

File: src/Main.py [hunk 16]
```diff
@@ -0,0 +1,678 @@
...
+        output = self.run_command(
+            'grep -e ":x:[0-9][0-9][0-9][0-9]:" {}/etc/passwd | cut -f 1 -d ":"'.format(TEMPDIR))
+        if output == "":
+            return []
+        self.run_command("umount -lf {}".format(TEMPDIR))
+        users = output.split()
+        users.append("root")
+        return users
+
+    def list_mbrs(self):
```
Comment (sulincix) [inline src/Main.py:605]: Read /sys/block/<mbr>/size file and if lower than 512 ignore.

empty sd card readers has mbr name but has no volume.

Also you dont need to run command for listing. Just use os.listdir()

File: src/Main.py [hunk 17]
```diff
@@ -0,0 +1,678 @@
...
+            return
+    def vte_exited(self, widget, status):
+        exit_status = os.waitstatus_to_exitcode(status)
+        if exit_status != 0:
+            self.update_status_page(_("An error occured"), "dialog-error-symbolic", _("An error occured while executing the command. Please check the logs"), True, True)
+            return
+        if self.post_command == None:
+            self.update_status_page(_("An error occured"), "dialog-error-symbolic", _("An error occurred after the command has been executed."), True, True)
+            return
+        Thread(target=self.post_command).start()
```
Comment (sulincix) [inline src/Main.py]: You can change automatically status page if log page shown.

File: src/Main.py [hunk 18]
```diff
@@ -0,0 +1,678 @@
...
+        # loading page
+        self.btn_go_mainpage = self.builder.get_object("button_loading_to_mainpage")
+        self.btn_close_logs = self.builder.get_object("button_close_logs")
+        self.status_page = self.builder.get_object("page_status")
+        self.spinner_loading = self.builder.get_object("spinner_page_loading")
+        self.btn_show_log = self.builder.get_object("button_show_log")
+        self.box_vte = self.builder.get_object("box_vte")
+
+        # Vte Terminal
+        self.vte_terminal = Vte.Terminal()
```
Comment (sulincix) [inline src/Main.py:64]: copy to clipboard required.

File: src/Main.py (hunk 15 above)
Comment (Vilez0) [reply src/Main.py]: done

File: src/Main.py (hunk 13 above)
Comment (Vilez0) [reply src/Main.py]: [8e17796](https://github.com/pardus/pardus-boot-repair/pull/3/commits/8e177969c31774401544767b5d0fbb6b39597745)

File: src/Main.py (hunk 16 above)
Comment (Vilez0) [reply src/Main.py:605]: ec54fa2

File: src/Main.py (hunk 17 above)
Comment (Vilez0) [reply src/Main.py]: done: 8c14a96

File: src/Main.py (hunk 14 above)
Comment (Vilez0) [reply src/Main.py]: 82c1511

File: src/Main.py (hunk 18 above)
Comment (Vilez0) [reply src/Main.py:64]: done 242522f

**Pull request merged**